src/Storages/MergeTree/MergeTreeRangeReader.cpp

```diff
@@ -348,74 +353,250 @@ void MergeTreeRangeReader::ReadResult::shrink(Columns & old_columns)
+    assert(!final_filter.present()
+        || final_filter.countBytesInFilter() == num_rows /// If filter has been applied.
+        || total_rows_per_granule == num_rows  /// If filter has not been applied.
+    );
+
+    /// Check that additional columns have the same number of rows as the main columns.
+    assert(!additional_columns || additional_columns.rows() == num_rows);
+
+    for (const auto & column : columns)
+    {
+        if (column)
+            assert(column->size() == num_rows);
+    }
+}
+
+std::string MergeTreeRangeReader::ReadResult::dumpInfo() const
+{
+    WriteBufferFromOwnString out;
+    out << "num_rows: " << num_rows
+        << ", columns: " << columns.size()
+        << ", total_rows_per_granule: " << total_rows_per_granule;
+    if (final_filter.present())
+    {
+        out << ", filter size: " << final_filter.size()
+        << ", filter 1s: " << final_filter.countBytesInFilter();
+    }
+    else
+    {
+        out << ", no filter";
+    }
+    for (size_t ci = 0; ci < columns.size(); ++ci)
+    {
+        out << ", column[" << ci << "]: ";
+        if (!columns[ci])
+            out << " nullptr";
+        else
+        {
+            out << " " << columns[ci]->dumpStructure();
+        }
+    }
+    if (additional_columns)
+    {
+        out << ", additional_columns: " << additional_columns.dumpStructure();
+    }
+    return out.str();
+}
+
+static std::string dumpNames(const NamesAndTypesList & columns)
+{
+    WriteBufferFromOwnString out;
+    for (auto it = columns.begin(); it != columns.end(); ++it)
+    {
+        if (it != columns.begin())
+            out << ", ";
+        out << it->name;
+    }
+    return out.str();
+}
+
 void MergeTreeRangeReader::ReadResult::setFilterConstTrue()
 {
-    clearFilter();
-    filter_holder = DataTypeUInt8().createColumnConst(num_rows, 1u);
+    /// Remove the filter, so newly read columns will not be filtered.
+    final_filter = FilterWithCachedCount();
 }
 
-void MergeTreeRangeReader::ReadResult::setFilterConstFalse()
+ColumnPtr andFilters(ColumnPtr c1, ColumnPtr c2)
 {
-    clearFilter();
-    columns.clear();
-    num_rows = 0;
+    // TODO: use proper vectorized implementation of AND?
+    assert(c1->size() == c2->size());
+    auto res = ColumnUInt8::create(c1->size());
+    auto & res_data = res->getData();
+    const auto & c1_data = typeid_cast<const ColumnUInt8&>(*c1).getData();
+    const auto & c2_data = typeid_cast<const ColumnUInt8&>(*c2).getData();
+    const size_t size = c1->size();
+    const size_t step = 16;
+    size_t i = 0;
+    /// NOTE: '&&' must be used instead of '&' for 'AND' operation because UInt8 columns might contain any non-zero
+    /// value for true and we cannot bitwise AND them to get the correct result.
+    for (; i + step < size; i += step)
+        for (size_t j = 0; j < step; ++j)
+            res_data[i+j] = (c1_data[i+j] && c2_data[i+j]);
```

Comment:
I hope this loop vectorizes well.
But actually, I think a simple loop would be vectorized in the same way.
What can we improve is actually to round up to 16 bytes, because ColumnUInt8 uses PaddedPodArray with 15 bytes padding (so we won't need a tail).